src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Amazon;
+using Amazon.S3;
+using Amazon.S3.Model;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Extensions;
+using Microsoft.Extensions.Options;
+using SixLabors.ImageSharp.Web.Resolvers;
+using SixLabors.ImageSharp.Web.Resolvers.AWS;
+
+namespace SixLabors.ImageSharp.Web.Providers.AWS
+{
+    /// <summary>
+    /// Returns images stored in AWS S3.
+    /// </summary>
+    public class AWSS3StorageImageProvider : IImageProvider
+    {
+        /// <summary>
+        /// Character array to remove from paths.
+        /// </summary>
+        private static readonly char[] SlashChars = { '\\', '/' };
+
+        /// <summary>
+        /// The containers for the blob services.
+        /// </summary>
+        private readonly Dictionary<string, AmazonS3Client> buckets
+            = new();
+
+        private readonly AWSS3StorageImageProviderOptions storageOptions;
+        private Func<HttpContext, bool> match;
+
+        /// <summary>
+        /// Contains various helper methods based on the current configuration.
+        /// </summary>
+        private readonly FormatUtilities formatUtilities;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AWSS3StorageImageProvider"/> class.
+        /// </summary>
+        /// <param name="storageOptions">The S3 storage options</param>
+        /// <param name="formatUtilities">Contains various format helper methods based on the current configuration.</param>
+        public AWSS3StorageImageProvider(IOptions<AWSS3StorageImageProviderOptions> storageOptions, FormatUtilities formatUtilities)
+        {
+            Guard.NotNull(storageOptions, nameof(storageOptions));
+
+            this.storageOptions = storageOptions.Value;
+
+            this.formatUtilities = formatUtilities;
+
+            foreach (AWSS3BucketClientOptions bucket in this.storageOptions.S3Buckets)
+            {
+                AmazonS3Client s3Client;
```

Comment:
We should probably use the same factory code in both the provider and cache for this.